{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script finds a human TAD for each human CNE and uses the homology between the TADs (generated by `compareTADsizes`) to generate a `lastz` job and search for the axolotl counterpart of the CNE in the respective axolotl TAD."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "opts_tads = ['/groups/tanaka/Projects/axolotl-genome/AmexG_v6.0/AmexG_v6.0_DD/work/manuscript/rebuttal/TADs_and_CNEs/homologous_TADs.hg19_vs_amex.list']\n",
    "opts_cnes = ['/groups/tanaka/Projects/axolotl-genome/AmexG_v6.0/AmexG_v6.0_DD/work/manuscript/rebuttal/TADs_and_CNEs/human_enhancers.hg19.fa']\n",
    "opts_out = ['/groups/tanaka/Projects/axolotl-genome/AmexG_v6.0/AmexG_v6.0_DD/work/manuscript/compare_TADs/homologous_TADs.hg19_vs_amex.with_CNEs.list']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import sys\n",
    "import re\n",
    "\n",
    "def loadCNEs(fastaFile):\n",
    "    data = []\n",
    "    with(open(fastaFile, 'r')) as hFile:\n",
    "        for line in hFile.readlines():\n",
    "            if not line.startswith('>'):\n",
    "                continue\n",
    "            chrName, start, end = line[1:].strip().split('_')\n",
    "            data.append([chrName, int(start), int(end)])\n",
    "    return pd.DataFrame(data=data, columns=['chr', 'start', 'end'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "cnes_hg = loadCNEs(opts_cnes[0])\n",
    "\n",
    "# For each TAD, find the list of CNEs it overlaps\n",
    "hOut = open(opts_out[0], 'w')\n",
    "with(open(opts_tads[0], 'r')) as hFile:\n",
    "    for line in hFile.readlines():\n",
    "        cols = line.strip().split('\\t')\n",
    "        m = re.search('^([^:]+):([0-9]+)-([0-9]+)$', cols[0])\n",
    "        if not m:\n",
    "            print(f\"ERROR: Malformatted line: '{line}'\", file=sys.stderr)\n",
    "\n",
    "        start = int(m.group(2))\n",
    "        end = int(m.group(3))\n",
    "        _cnes = cnes_hg[(cnes_hg['chr'] == m.group(1)) & (cnes_hg['start'] >= start) & (cnes_hg['end'] <= end)]\n",
    "        if _cnes['chr'].count() == 0:\n",
    "            cols.insert(1, 'N/A')\n",
    "        else:\n",
    "            cols.insert(1, str([f'{x[0]}:{x[1]}-{x[2]}' for x in _cnes.values]))\n",
    "        print(\"\\t\".join(cols), file=hOut)\n",
    "hOut.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    chr     start       end\n",
      "0  chr1   3190581   3191428\n",
      "1  chr1   8130439   8131887\n",
      "2  chr1  10593123  10594209\n",
      "3  chr1  10732070  10733118\n",
      "4  chr1  10757664  10758631\n"
     ]
    }
   ],
   "source": [
    "print(cnes_hg.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
